### PR TITLE
docs: add CLI doc

### PR DESCRIPTION
**Pull request by Jcahilltorre**
Adds the doc for the unitctl cli

To do:
- Add the links to the binaries or remove the section
- Decide whether to add the compiling instructions here or keep them as a link to the repo

### CODE REVIEW

**Comment by avahahn** [comment]
I see that we have a global section for options. I am curious if there should perhaps be a section for options per subcommand?

**Review by avahahn — CHANGES_REQUESTED** [review]
I see that we have a global section for options. I am curious if there should perhaps be a section for options per subcommand?

I also note that there is a subcommand missing (export).

**Comment by avahahn** [comment]
> Decide whether to add the compiling instructions here or keep them as a link to the repo

My vote is to link to the repo for this. The instructions may change along with planned improvements to the unitctl design. If we link to the repo unitctl readme then readers will always be pointed to the most up to date instructions.

**Comment by avahahn** [comment]
> Add the links to the binaries or remove the section

I believe a good option here is to link to the github releases section in the Unit repo, as that will populate with download links for the CLI.

**Comment by Jcahilltorre** [comment]
> > Add the links to the binaries or remove the section
> 
> I believe a good option here is to link to the github releases section in the Unit repo, as that will populate with download links for the CLI.

I don't see a releases section in the Unit repo currently

**Comment by Jcahilltorre** [comment]
> I see that we have a global section for options. I am curious if there should perhaps be a section for options per subcommand?
> 
> I also note that there is a subcommand missing (export).

the list of command and options is based on the documentation in the repo. The export command is missing there too. Where could I get a comprehensive list of the options for each command?

File: source/unitctl.rst [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+interface (unitctl) that exposes the functionality provided by the SDK.
+
+This CLI is a powerful tool that allows you to deploy, manage, and configure
+Unit in your environment.
+
+*****************
+Download binaries
+*****************
+
+We provide unitctl binaries for Linux (both ARM64 and X64) and MacOS systems.
```
Comment (travisamartin) [inline source/unitctl.rst:21]:
```suggestion
Unitctl binaries are available for Linux (ARM64 and X64) and MacOS systems.
```

File: source/unitctl.rst [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+You can find the latest binaries on the `Unit GitHub releases page
+<https://github.com/nginx/unit/releases>`_
```
Comment (travisamartin) [inline source/unitctl.rst:24]:
```suggestion
Download the latest binaries from the `Unit GitHub releases page <https://github.com/nginx/unit/releases>`_.
```

File: source/unitctl.rst [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+*****************
+Build from source
+*****************
+
+If you would like to build unitctl from source, you can do so by following the
+instructions in the `unitctl repository
+<https://github.com/nginx/unit/tree/master/tools/unitctl>`_.
```
Comment (travisamartin) [inline source/unitctl.rst:32]:
```suggestion
To build unitctl from source, follow the instructions in the `unitctl repository <https://github.com/nginx/unit/tree/master/tools/unitctl>`_.
```

File: source/unitctl.rst [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+*************
+Using unitctl
+*************
+
+The unitctl CLI provides a number of commands that allow you to interact with
+Unit. The following is a list of the available commands:
```
Comment (travisamartin) [inline source/unitctl.rst:39]:
```suggestion
The unitctl CLI offers several commands to interact with Unit. Here are the available commands:
```

File: source/unitctl.rst [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+       tarball
+
+   * - **import**
+     - Import Unit configuration from a directory
+
+   * - **execute**
+     - Send a raw JSON payload to Unit
+
+   * - **status**
+     - Get the current status of the Unit
```
Comment (travisamartin) [inline source/unitctl.rst:67]:
```suggestion
     - Get the current status of Unit
```

File: source/unitctl.rst [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+   * - **-v, --version**
+     - Display the version of the unitctl CLI
+
++++++++++++++++++++++++++++++++++
+List and create instances of Unit
++++++++++++++++++++++++++++++++++
+
+The **instances** command allows you to list all running Unit processes and
+deploy new instances of Unit.
```
Comment (travisamartin) [inline source/unitctl.rst:105]:
```suggestion
The **instances** command lets you list all running Unit processes and
deploy new instances of Unit.
```

File: source/unitctl.rst [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+The **instances** command has the following options:
```
Comment (travisamartin) [inline source/unitctl.rst:107]:
```suggestion
The **instances** command has the following option:
```

File: source/unitctl.rst [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+   :header-rows: 1
+
+   * - Option
+     - Description
+
+   * - **new**
+     - Deploy a new instance of Unit
+
+Running unitcl with the **instances** command will display an output similar to
+the following:
```
Comment (travisamartin) [inline source/unitctl.rst:119]:
```suggestion
Running unitcl with the **instances** command shows output similar to this:
```

File: source/unitctl.rst (hunk 5 above)
Comment (Jcahilltorre) [reply source/unitctl.rst:67]: this was a leftover from a previous version of the sentence that mentioned the Unit instance itselv

File: source/unitctl.rst [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+   No socket path provided - attempting to detect from running instance
+   unitd instance [pid: 79489, version: 1.32.0]:
+      Executable: /opt/unit/sbin/unitd
+      API control unix socket: unix:/opt/unit/control.unit.sock
+      Child processes ids: 79489, 79489
+      Runtime flags: --no-daemon
+      Configure options: --prefix=/opt/unit --user=myUser --group=myGroup --openssl
+
+The **instances** command can also be used to deploy a new instance of Unit
+using the **new** option and three arguments:
```
Comment (travisamartin) [inline source/unitctl.rst:133]:
```suggestion
You can use the **new** option with three arguments to deploy a new instance of Unit:
```

File: source/unitctl.rst [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+
+1. A means to show the control API: Either a file path to open a unix socket, or
+   a TCP address with port.
```
Comment (travisamartin) [inline source/unitctl.rst:130]:
```suggestion
1. **Control API path**: A file path for a Unix socket or a TCP address with port.
```

File: source/unitctl.rst [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+     from the host. For example: **/tmp/2**.
+   - If a TCP address is specified, the Unit container will listen on this
+     address and port. For example: **127.0.0.1:7171**.
+
+2. A path to an application. The Unit container will mount this in READ ONLY mode
+   to **/www** internally. This will allow the user to configure their Unit
+   container to expose an application stored on the host. For example: **$(pwd)**.
+
+3. An image tag. Unitctl will deploy this image, enabling you to deploy custom
+   images. For example: **unit:wasm**.
```
Comment (travisamartin) [inline source/unitctl.rst:143]:
```suggestion
   - if you specify a directory, the Unit container will mount it to **/var/run** internally. 
     The control socket and pid file are accessible from the host. Example: **/tmp/2**.
   - If you specify a TCP address, the Unit container will listen on this
     address and port. Example: **127.0.0.1:7171**.

2. **Application path**. The Unit container will mount this path in read-only mode
   to **/www** internally. This setup allows you to configure the Unit
   container to expose an application stored on the host. Example: **$(pwd)**.

3. **Image tag**: Unitctl will deploy this image, enabling you use custom
   images. For example: **unit:wasm**.
```

File: source/unitctl.rst [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+This CLI is a powerful tool that allows you to deploy, manage, and configure
+Unit in your environment.
+
+*****************
+Download binaries
+*****************
+
+Unitctl binaries are available for Linux (ARM64 and X64) and MacOS systems.
+
+Download the latest binaries from the `Unit GitHub releases page <https://github.com/nginx/unit/releases>`_.
```
Comment (Jcahilltorre) [inline source/unitctl.rst:23]:
```suggestion
Download the latest binaries from the `Unit GitHub releases page
<https://github.com/nginx/unit/releases>`_.
```

File: source/unitctl.rst [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+
+*****************
+Build from source
+*****************
+
+To build unitctl from source, follow the instructions in the `unitctl repository <https://github.com/nginx/unit/tree/master/tools/unitctl>`_.
```
Comment (Jcahilltorre) [inline source/unitctl.rst:29]:
```suggestion
To build unitctl from source, follow the instructions in the `unitctl repository
<https://github.com/nginx/unit/tree/master/tools/unitctl>`_.
```

File: source/unitctl.rst [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+   }
+
+.. note::
+
+   This command supports operating on multiple instances of Unit at once. To do
+   this, use the **-s** option multiple times with different values:
+
+   .. code-block:: console
+
+      $ unitctl -s '127.0.0.1:8001' -s /run/nginx-unit.control.sock app list
```
Comment (travisamartin) [inline source/unitctl.rst:204]:
```suggestion
+++++++++++++++++++++++++++++
List and restart running apps
+++++++++++++++++++++++++++++

The **app** command lets you list and restart active applications.

Options
-------

The **app** command has the following options:

.. list-table::
   :header-rows: 1

   * - Option
     - Description

   * - **list**
     - List all active applications

   * - **reload <APP_NAME>**
     - Restart the specified application

To list active applications, run:

.. code-block:: console

   $ unitctl app list
   {
     "wasm": {
        "type": "wasm-wasi-component",
        "component": "/www/wasmapp-proxy-component.wasm"
     }
   }

To restart an application, run:

.. code-block:: console

   $ unitctl app reload wasm
   {
      "success": "Ok"
   }

.. note::

   This command supports operating on multiple instances of Unit at once. To do
   this, use the **-s** option multiple times with different values:

   .. code-block:: console

      $ unitctl -s '127.0.0.1:8001' -s /run/nginx-unit.control.sock app list
```

Comment (Jcahilltorre) [reply source/unitctl.rst:204]: The instances command also has an (optional) option in a table. I feel if we apply this change we also should have an Options header in the previous command for consistency

File: source/unitctl.rst [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+
+++++++++++++++++++++++
+Fetch active listeners
+++++++++++++++++++++++
+
+Unitctl can query a given control API to fetch all configured listeners.
+
+To list all active listeners, run the following command:
```
Comment (travisamartin) [inline source/unitctl.rst:212]:
```suggestion
To list all active listeners, run:
```

File: source/unitctl.rst (hunk 14 above)
Comment (Jcahilltorre) [reply source/unitctl.rst:204]: in L105

File: source/unitctl.rst [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+      $ unitctl -s '127.0.0.1:8001' -s /run/nginx-unit.control.sock listeners
+
+++++++++++++++++++++++++
+Check the status of Unit
+++++++++++++++++++++++++
+
+Unitctl can query the control API to provide the **status** of the running Unit
+daemon.
+
+To get the current status of the Unit, run the following command:
```
Comment (travisamartin) [inline source/unitctl.rst:240]:
```suggestion
To get the current status of the Unit, run:
```

File: source/unitctl.rst [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+      $ unitctl -s '127.0.0.1:8001' -s /run/nginx-unit.control.sock status
+
++++++++++++++++++++++++++++++++++++
+Send configuration payloads to Unit
++++++++++++++++++++++++++++++++++++
+
+Using the **execute** command, Unitctl can accept custom request payloads and
+query given API endpoints with them. The request payload must be passed in using
+the **-f** flag either as a filename or using the **-** filename to denote the
+use of stdin as shown in the example below.
```
Comment (travisamartin) [inline source/unitctl.rst:271]:
```suggestion
With the **execute** command, Unitctl can accept custom request payloads and
query specified API endpoints with them. Use the **-f** flag to pass the request 
payload as a filename or **-** to denote stdin, as shown in the example below.
```

File: source/unitctl.rst (hunk 14 above)
Comment (travisamartin) [reply source/unitctl.rst:204]: Agreed, both should have the header. I just learned how to add headers in rst -- this is my first one. I can add the other.

File: source/unitctl.rst [hunk 18]
```diff
@@ -0,0 +1,388 @@
...
+interface (unitctl) that exposes the functionality provided by the SDK.
+
+This CLI is a powerful tool that allows you to deploy, manage, and configure
+Unit in your environment.
+
+*****************
+Download binaries
+*****************
+
+Unitctl binaries are available for Linux (ARM64 and X64) and MacOS systems.
```
Comment (ac000) [inline source/unitctl.rst:21]: MacOS should be spelt 'macOS'.

Comment (Jcahilltorre) [reply source/unitctl.rst:21]:
```suggestion
Unitctl binaries are available for Linux (ARM64 and X64) and macOS systems.
```

File: source/unitctl.rst [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+
+.. code-block:: console
+
+   {
+   "success": "Reconfiguration done."
+   }
+
+.. note::
+
+   This command does not support operating on multiple instances of Unit at once.
```
Comment (travisamartin) [inline source/unitctl.rst:333]:
```suggestion
Unitctl can fetch the configuration from a running instance of Unit and load it
in a preconfigured editor on your command line using the **edit** command.

Unitctl tries to use the editor configured with the **EDITOR** environment
variable, but defaults to vim, emacs, nano, vi, or pico if **EDITOR** is not set.

To edit the current configuration, run:

.. code-block:: console

   $ unitctl edit

The configuration loads into the editor, allowing you to make any necessary
changes. Once you save and close the editor, you see the following output:

.. code-block:: console

   {
   "success": "Reconfiguration done."
   }

.. note::

   This command does not support operating on multiple instances of Unit at once.
```

File: source/unitctl.rst [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+
++++++++++++++++++++++++++++++++++++++++++
+Importing the configuration from a folder
++++++++++++++++++++++++++++++++++++++++++
+
+Using the **import** command, Unitctl will parse existing configuration,
+certificates, and NJS modules stored in a directory and convert them into a
+payload to reconfigure a given Unit daemon.
+
+To export the configuration, run the following command:
```
Comment (travisamartin) [inline source/unitctl.rst:343]:
```suggestion
The **import** command lets Unitctl read configuration files, certificates, and
NJS modules from a directory. Unitctl then converts these files into a payload
to reconfigure a Unit daemon.

To export the configuration, run:
```

File: source/unitctl.rst [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+
++++++++++++++++++++++++++++++++++++++
+Exporting the configuration from Unit
++++++++++++++++++++++++++++++++++++++
+
+The **export** command will query a control API to fetch running configuration
+and NJS modules from a Unit process. Due to a technical limitation this output
+will not contain the currently stored certificate bundles. The output is saved
+as a tarball at the filename given with the **-f** argument. Standard out may be
+used with **-f -** as shown in the following examples:
```
Comment (travisamartin) [inline source/unitctl.rst:361]:
```suggestion
The **export** command queries a control API to fetch the running configuration
and NJS modules from a Unit process. The output does not include the currently
stored certificate bundles due to a technical limitation. The output is saved
as a tarball with the filename specified by the **-f** argument. You can also
use standard output with **-f -**, as shown in the examples below:
```

File: source/unitctl.rst [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+
+.. note::
+
+   This command does not support operating on multiple instances of Unit at once.
+
++++++++++++++++++++++++++++++++++
+Wait for a socket to be available
++++++++++++++++++++++++++++++++++
+
+All commands support waiting on unix sockets for availability:
```
Comment (travisamartin) [inline source/unitctl.rst:382]:
```suggestion
All commands support waiting for Unix sockets to become available:
```

**Comment by avahahn** [comment]
> > > Add the links to the binaries or remove the section
> > 
> > 
> > I believe a good option here is to link to the github releases section in the Unit repo, as that will populate with download links for the CLI.
> 
> I don't see a releases section in the Unit repo currently

https://github.com/nginx/unit/releases
The section is there, and will be populated with the upcoming 1.33 release

**Comment by avahahn** [comment]
> > I see that we have a global section for options. I am curious if there should perhaps be a section for options per subcommand?
> > I also note that there is a subcommand missing (export).
> 
> the list of command and options is based on the documentation in the repo. The export command is missing there too. Where could I get a comprehensive list of the options for each command?

https://github.com/nginx/unit/tree/master/tools/unitctl#export-configuration-from-a-running-unit-instance

It is listed here.

**Review by avahahn — APPROVED** [review]

**Review by travisamartin — APPROVED** [review]

**Pull request merged**